R/assocTestAggregate.R

```diff
@@ -182,99 +133,116 @@ setMethod("assocTestAggregate",
-                                              # pval.method=pval.method)
-                      res[[i]] <- cbind(res[[i]], assoc, stringsAsFactors=FALSE)
-                  }
-
-                  if (verbose & n.iter > 1 & i %% set.messages == 0) {
-                      message(paste("Iteration", i , "of", n.iter, "completed"))
+                  } else {
+                      weight <- NULL
                   }
-                  i <- i + 1
-                  iterate <- GWASTools::iterateFilter(gdsobj)
+                  
+                  return(list(var.info=var.info, geno=geno, chr=var.info$chr, weight=weight))
               }
-
-              res <- list(results=as.data.frame(rbindlist(res, use.names=TRUE, fill=TRUE)),
+              
+              
+              res <- bpiterate(ITER, .testGenoBlockAggregate, BPPARAM=BPPARAM,
+                               sex=sex, null.model=null.model, AF.max=AF.max,
+                               weight.beta=weight.beta, test=test, 
+                               neig=neig, ntrace=ntrace, rho=rho,
+                               sparse=FALSE, imputed=FALSE,
+                               male.diploid=male.diploid)
+              .stopOnError(res)
+
+              res1 <- lapply(res, function(x) x[[1]])
+              res.var <- lapply(res, function(x) x[[2]])
+              rm(res)
+              res <- list(results=as.data.frame(rbindlist(res1, use.names=TRUE, fill=TRUE)),
                           variantInfo=res.var)
               .annotateAssoc(gdsobj, res)
           })
+
+
+.testGenoBlockAggregate <- function(x, sex, null.model, AF.max,
+                                    weight.beta, test, 
+                                    neig, ntrace, rho,
+                                    sparse, imputed, male.diploid, ...) {
+    
+    x <- .prepGenoBlock(x, AF.max=AF.max, sex=sex, imputed=imputed, male.diploid=male.diploid)
+    var.info <- x$var.info
+    n.obs <- x$n.obs
+    freq <- x$freq
+    geno <- x$geno
+    weight <- x$weight
+    rm(x)
+    
+    # weights
+    if (is.null(weight)) {
+        # Beta weights
+        if (test %in% c("BinomiRare", "CMP")){
+            weight <- seq(from=1, to=1, length.out=length(freq$freq))
+        } else {
+            weight <- .weightFromFreq(freq$freq, weight.beta)
+        }
+    } else {
+        # user supplied weights
+        weight0 <- is.na(weight) | weight == 0
+        if (any(weight0)) {
+            keep <- !weight0
+            var.info <- var.info[keep,,drop=FALSE]
+            geno <- geno[,keep,drop=FALSE]
+            n.obs <- n.obs[keep]
+            freq <- freq[keep,,drop=FALSE]
+            weight <- weight[keep]
+        }
+    }
+    
+    # number of variant sites
+    n.site <- length(unique(var.info$variant.id))
+    
+    # number of alternate alleles
+    n.alt <- sum(geno, na.rm=TRUE)
+    
+    # number of samples with observed alternate alleles > 0
+    n.sample.alt <- sum(rowSums(geno, na.rm=TRUE) > 0)
+    
```

Comment:
since we're imputing first, we have to change the way we get `n.sample.alt`. this one is a bit trickier, but here's what I came up with:
`n.sample.alt <- sum(.apply(geno, 1, function(x){ any(x >= 1) }))`

Basically, check by row (sample) if any values are at least 1. I'm assuming that any imputed value would be < 1. This assumption will hold as long as the MAF != 0.5, which would be unlikely for rare variant tests.